### PR TITLE
Final Developer's bible

### LINKED ISSUE #54: Follow-up: Bible for Developers
**Objectives**

To integrate everyone's feedback on each other's work on the Bible for Developers.

**Note that reviews should be conducted within the *Developers-Bible branch*, and include the following pages:**
  * Bible for Developers
  * General Items Breakdown
  - Stage 1 Breakdown
  - Stage 2 Breakdown
  - Stage 3 Breakdown
  - Stage 4 Breakdown
  * When making a comment, please find the appropriate page on GitHub's review (under "Files Changed" within the pull request menu)

- [ ] Everyone provides initial feedback review of everyone's work on #68 
  - [x] Review from @Jmendoza194 
  - [x] Review from @joeyhuaa 
  - [ ] Review from @julma1040 
  - [x] Review from @atuljayaram 

- [ ] Organize meeting on Saturday to discuss entire Bible

- [ ] Follow-up work done on the following: (from #30 )

- [ ] General Items Documentation 
- [ ] Stage 1 Documentation (mostly updating)
- [ ] Stage 2 Documentation (mostly updating)
  - [ ] Checkpoints need to be documented
  - [ ] Concepts need to be documented
- [x] Stage 3 Documentation (creating from scratch using READMEs made by @wongband (Bryan) as inspiration)
  - [x] READMEs should be edited to be more clear
  - [x] Documentation for Test Cases should be edited to be more clear
- [ ] Stage 4 Documentation (creating from scratch)
- [ ] How this Checklist should be exactly used by Daniel (President), Kevin (or any Head of DevRel), managers and developers

**Collaborators and Tasks**
@kavuong - initial review of everyone's work, organize meeting
All collaborators from #30:  
@joeyhuaa - Stage 1 Documentation + Stage 2 Documentation + General Steps Documentation
@atuljayaram - Stage 3 Documentation
@julma1040 - Stage 4 Documentation
@Jmendoza194 - how this checklist should be used by everyone

**Deadline**
Wednesday, 3/31 initial review by everyone 
Follow-up work by everyone by Friday, 4/3

Final Review by @kavuong Wed, 4/8

### PR DESCRIPTION
**Pull request by kavuong**
Hi everyone: please provide a review on the General Stage Breakdown and the Stage 1-4 Breakdown pages. (Gitbooks updates every branch every time a commit is made to master for some reason)

the files to review are labelled as:
* teams/developer-relations/bible-for-developers/general-items-breakdown.md 
* teams/developer-relations/bible-for-developers/stage-1-breakdown.md 
  * (same for stages 2-4)

you can find it easier by filtering by the following commit:
name: "GitBook: [developers-bible] 76 pages modified"
commit ID: 2d8978d
@vkxu657 please review and edit this when they have finished editing and providing their input (I will let you know).

#### Issues Closed 
_Please make one new line for each issue, otherwise not all issues will be accounted for!_

- [ ] Closes #54 

...

Changes proposed in this pull request:
* Bible for Developers Finalized

### CODE REVIEW

File: teams/developer-relations/bible-for-developers/general-items-breakdown.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+In the above example, **1-1-1.md is the first easy card associated with the first medium card associated with the first hard card.** Can you translate 2-1-1.md into plain English?
+
+#### Card Length
+
+Cards should be concise. Ideally, you should not have to scroll more than once to view the whole card. Stuffing a ton of information in a single card can be too overwhelming, so try to keep things brief and split up topics as much as possible.
+
+#### Writing Conventions
+
+Last but certainly not least, proper writing conventions such as grammar, punctuation, capitalization, etc. should be utilized.
+
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/general-items-breakdown.md:51]: Everything looks good here!

File: teams/developer-relations/bible-for-developers/stage-1-breakdown.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+### Stages
+
+Here's an in-depth breakdown of how cards progress in the BitProject system.
+
+#### Stage 1 - Starting Content
+
+**Card content should be accessible to beginners**, meaning someone with no prior knowledge should be able to follow the content. This goes hand-in-hand with using a conversational writing style.
+
+Next up is what we call the **Micro to Macro Principle**. For labs and activities, please start at a micro level and then move up. For example, students should not code the main\(\) function first. As a rule of thumb, you should never have to utter the words “this function will be coded in a later part”. Every hard card should work on its own and not have to rely on later work for it to work out.
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-1-breakdown.md:11]: This paragraph could be worded a bit better. Some awkward phrasing.

File: teams/developer-relations/bible-for-developers/stage-1-breakdown.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+![Micro &#x2014;&amp;gt; Macro!](../../../.gitbook/assets/smalltobig.jpeg)
+
+**Images**! Make liberal use of images \(like the one above\) in your cards. It's much easier to visualize a concept than to read three paragraphs of text. Obviously, text is necessary, but pairing text with visuals is immensely helpful. Also, images help break up blocks of text, making it easier on the eyes. Please use copyright-free images, Pexels is a great source. Make sure images are embedded using &lt;img&gt; tags and not URLs.
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-1-breakdown.md:15]: Some repetition here. "easier to visualize concepts" AND "easier on the eyes"

File: teams/developer-relations/bible-for-developers/stage-1-breakdown.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+When it comes to Labs, there are a number of things to keep in mind.
+
+**Hard cards** are, by definition, _hard._ A hard card for the Sudoku Solver Lab might explain the game of Sudoku and have a teeny-tiny bit of **starter code** at the end. Keep in mind that, although this card should stretch the student's brain, it should also provide just enough guidance for the student to finish the task. The main thing to keep in mind is you _should NOT_ leak any hints_._ Example [here](https://github.com/bitprj/curriculum/blob/master/Data-Structures-and-Algos-Topic/labs/Lab6_Sudoku_Solver/Cards/1.md).
+
+**Medium cards** are probably the most difficult cards to make, since you have to find a delicate middle ground between hard and easy. That said, keep in mind that the goal of medium is to serve as a bridge between hard and easy. You start to squeeze out a few hint droplets into the student's cup, but don't fill the entire cup. Example [here](https://github.com/bitprj/curriculum/blob/master/Data-Structures-and-Algos-Topic/labs/Lab6_Sudoku_Solver/Cards/11.md).
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-1-breakdown.md:38]: Repetition in the first two sentences. Also, "middle ground between hard and easy" isn't that informative on its own. Replace "hard" and "easy" with what those cards are meant to accomplish.

File: teams/developer-relations/bible-for-developers/stage-2-breakdown.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+Each Activity/Lab should contain a brief **summary** describing it using 1 sentence only. In addition, please provide a **long summary** that expands on the main concepts covered in the Activity/Lab, about 3-4 sentences total. Here's an example:
+
+![Example of an Activity summary.](../../../.gitbook/assets/screen-shot-2020-03-30-at-1.40.58-pm.png)
+
+#### Stage 2 - Lab Addendum
+
+For each lab, **test cases** must be provided. Make sure your lab is concise enough so that test cases can be made. For example, if you were making a Wheel of Fortune lab, don’t ask users to make their own lists of possible messages. Instead, create the messages yourself and have them simply make a list out of those predefined messages. In the scenario where students get to create their own list, it would be impossible to actually define test cases for each student’s custom list.
+
+For the test cases, simply provide a document detailing what should be inputted and its output. There should be around 8-10 test cases per Lab.
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-2-breakdown.md:23]: @kavuong Doesn't this information overlap with Stage 3 a bit?

File: teams/developer-relations/bible-for-developers/stage-3-breakdown.md [hunk 6]
```diff
@@ -0,0 +1,934 @@
+# Stage 3 Breakdown
+
+
+
+### Stage 3 Documentation \(Atul\)
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-3-breakdown.md:5]: Get rid of this *me*.

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+![](https://lh6.googleusercontent.com/qvvB1JPKR_jm5GzbY8MyJtZ5XfJkzJiOsZTS6KswZLK_ivUf5REWuELLhDuImuoqxbSveiiCCtG8ok63KRArxcv28cCQC8ROfS2d2jkTB-9RlT7aCPOdmp95tXau3xAebj1o8VDy)
+
+You see how each module’s names are formatted Module\# name. Please do only use dashes \(-\) between words as any colons, semicolons, and slashes \(: ; / \) will mess with Github and cause trouble.
+
+Then into Module1-Intro-to-Data-Structures-and-Algos
+
+![](https://lh5.googleusercontent.com/b2jiQTcHdt07reu4h8trpAbq57cAEV5UpZfxhrImoP_a-f5mhYcUwCupfMSdX46RcXmWSZoJeQG_Ld7yauBa1ULvIhHKxcZcD5zdfn5si2XZBpvRQwkNyMV7JEn8P2-ZCNDwzQ8e)
+
+Again, you see how activite’s names are formatted with Activity\# name and it’s title in one chunk but capitalized every first letter of each word. Please do only use dashes \(-\) between words as any colons, semicolons, and slashes \(: ; / \) will mess with Github and cause trouble.
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-4-breakdown.md:25]: Spelling mistake, also the first sentence is worded awkwardly.

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+![](https://lh4.googleusercontent.com/Axm24Rp0aiNJ4PDkF9uyeIEJgFu08o2vam--oonqprZkYsvj-9DE4giGCxczXcd74hUvdtkKpyoQbwn05zOXRmK0DM4xBYsdEMY61mR9saD0Ex2mtOsWyVHacFiCgRtzjYEFEeWK)
+
+You can already see the way that this activity is being organized. Cards, checkpoints, and images goes in their own folder. Check out each of the folders and see how things are named.
+
+Cards are named from 1.md to \#.md.
+
+Checkpoints are named card\#-checkpoint.md.
+
+Images are named what the image is related to.
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-4-breakdown.md:41]: Could phrase this a little better.

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+Notice, the most important here is the README.md. The format of readmes have been documented here [https://about.bitproject.org/teams/engineering/readme-formats](https://about.bitproject.org/teams/engineering/readme-formats). Please follow the documentation and use the example in Github as a guide. Make sure you have all the information.
+
+\*NOTE: Gems will be explained in the next step so don’t worry about it yet.
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-4-breakdown.md:45]: I feel like this is a little unnecessary as you _literally_ explain gems in the very next line.

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+* [ ] **Gems**
+
+  Gems are something that students will receive upon completing a set of activities. Due to the difference in difficulty for each set of activities across all the modules, a set amount of gems does not make sense. To solve this issue, one of our members had created a Python script that you can run to calculate the number of gems that you should record in the readme of each activity.
+
+The idea is that if you per-assign a certain number of gems to each hard card and you know the amount of mediums and easy's under that given hard card, you can use the algorithm to calculate the entirety of the gem amounts for the rest of the cards.
+
+The script is written to assign points based on the types of cards. It’s in Python so make sure you have some sort of Python editor available for use. Download and open to follow along:
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-4-breakdown.md:53]: The last paragraph repeats the first paragraph a little. Also, some of the wording is a little awkward. Could be made more clear.

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+* [ ] **Gems**
+
+  Gems are something that students will receive upon completing a set of activities. Due to the difference in difficulty for each set of activities across all the modules, a set amount of gems does not make sense. To solve this issue, one of our members had created a Python script that you can run to calculate the number of gems that you should record in the readme of each activity.
+
+The idea is that if you per-assign a certain number of gems to each hard card and you know the amount of mediums and easy's under that given hard card, you can use the algorithm to calculate the entirety of the gem amounts for the rest of the cards.
+
+The script is written to assign points based on the types of cards. It’s in Python so make sure you have some sort of Python editor available for use. Download and open to follow along:  
+Calculates the amount of exp to subtract from medium
+
+**and easy cards for a single hard card \(NOT the entire lab\)**
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-4-breakdown.md:56]: Was this formatting intentional? If it is, nevermind.

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+_Step 4:_ With the results, please consult Kevin for the number of gems.
+
+* [ ] **Proofreading by the Writing Team**
+
+  This is the very last step of development. Please send your contents to the writing team \(currently Victoria Xu @@vkxu657 on Github\).
+
+This is important to make sure all the content is syntactically and grammatically correct so students won’t have trouble with the material for maximum understanding.
+
+It is your responsibility as a manager to raise any issues that are reported and issued to a member of your team to fix it by a deadline. All the writer’s changes should be integrated into your weekly PRs.
```
Comment (atuljayaram) [inline teams/developer-relations/bible-for-developers/stage-4-breakdown.md:114]: Could be worded a little better.

**Review by atuljayaram — CHANGES_REQUESTED** [review]

File: teams/developer-relations/bible-for-developers/general-items-breakdown.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+It's crucial that each card is numbered correctly so developers can keep lessons organized and students can follow a logical progression. BitProject uses two numbering schemes, one for Activities and one for Labs.
+
+Here's an example of an Activity \(pretty self-explanatory\):
+
+![Numbering for Activity Cards](../../../.gitbook/assets/screen-shot-2020-03-29-at-3.00.05-pm.png)
+
+And here's an example of a Lab:
+
+![Numbering for Lab Cards](../../../.gitbook/assets/screen-shot-2020-03-29-at-3.06.54-pm.png)
```
Comment (Jmendoza194) [inline teams/developer-relations/bible-for-developers/general-items-breakdown.md:32]: This is just me being nitpicky, But I think it might be better to show the example after you explained the numbering scheme. Also, I don't think the self-explanatory in parenthesis part is necessary. I realize you meant it for the activity cards, cuz it is self explanatory for that part. But I initially misunderstood that you meant that it was self-explanatory for the lab cards as well(which it isn't lol), and this could confuse beginners.

File: teams/developer-relations/bible-for-developers/stage-1-breakdown.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+Next up is what we call the **Micro to Macro Principle**. For labs and activities, please start at a micro level and then move up. For example, students should not code the main\(\) function first. As a rule of thumb, you should never have to utter the words “this function will be coded in a later part”. Every hard card should work on its own and not have to rely on later work for it to work out.
+
+![Micro &#x2014;&amp;gt; Macro!](../../../.gitbook/assets/smalltobig.jpeg)
+
+**Images**! Make liberal use of images \(like the one above\) in your cards. It's much easier to visualize a concept than to read three paragraphs of text. Obviously, text is necessary, but pairing text with visuals is immensely helpful. Also, images help break up blocks of text, making it easier on the eyes. Please use copyright-free images, Pexels is a great source. Make sure images are embedded using &lt;img&gt; tags and not URLs.
+
+Activities and Labs will feature coding tasks for students to complete. For these, developers should implement what we call **checkpoints**—Python console output that serves as a frame of reference for students to make sure their code is doing what it's supposed to be doing.
```
Comment (Jmendoza194) [inline teams/developer-relations/bible-for-developers/stage-1-breakdown.md:17]: The second sentence sounds awkward to me :/
Maybe rephrase it better?

File: teams/developer-relations/bible-for-developers/stage-1-breakdown.md [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+#### Stage 1 - Lab Addendum
+
+![Not a paid endorsement, this picture just fits perfectly.](../../../.gitbook/assets/inthelab.jpg)
+
+When creating a Lab, you are essentially giving the student a problem to solve, breaking it down into steps, with each step subdivided into cards of hard, medium, and easy difficulty. An example might be to code a Python program that can solve Sudoku.
+
+When it comes to Labs, there are a number of things to keep in mind.
+
+**Hard cards** are, by definition, _hard._ A hard card for the Sudoku Solver Lab might explain the game of Sudoku and have a teeny-tiny bit of **starter code** at the end. Keep in mind that, although this card should stretch the student's brain, it should also provide just enough guidance for the student to finish the task. The main thing to keep in mind is you _should NOT_ leak any hints_._ Example [here](https://github.com/bitprj/curriculum/blob/master/Data-Structures-and-Algos-Topic/labs/Lab6_Sudoku_Solver/Cards/1.md).
```
Comment (Jmendoza194) [inline teams/developer-relations/bible-for-developers/stage-1-breakdown.md:36]: "leak any hints_._"

Idk if you realize, but in the actual text, "hints_._" is what appears(At least thats what I see in the rich diff button).

File: teams/developer-relations/bible-for-developers/stage-3-breakdown.md [hunk 16]
```diff
@@ -0,0 +1,934 @@
...
+6 XXXXXXXXXX 6
+7 XXXXXXXXXX 7
+8 XXXXXXXXXX 8
+9 XXXXXXXXXX 9
+  0123456789
+Uh oh! You blew up!
+```
+
+> Note: Each test case file should contain 5-10 unique test cases.
+
```
Comment (Jmendoza194) [inline teams/developer-relations/bible-for-developers/stage-3-breakdown.md:934]: I think stage 3 breakdown is done well 👍

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+Walk through example:
+
+On the Bit curriculum Github, there is a module you should use as an example when trying to structure your module. Follow along by going to [https://bitproject.org/curriculum](https://bitproject.org/curriculum).
+
+Once at the link it should look like this. We’re going to go into the Data-Structures-and-Algos-Topic.
+
+![](https://lh6.googleusercontent.com/qvvB1JPKR_jm5GzbY8MyJtZ5XfJkzJiOsZTS6KswZLK_ivUf5REWuELLhDuImuoqxbSveiiCCtG8ok63KRArxcv28cCQC8ROfS2d2jkTB-9RlT7aCPOdmp95tXau3xAebj1o8VDy)
+
+You see how each module’s names are formatted Module\# name. Please do only use dashes \(-\) between words as any colons, semicolons, and slashes \(: ; / \) will mess with Github and cause trouble.
```
Comment (Jmendoza194) [inline teams/developer-relations/bible-for-developers/stage-4-breakdown.md:19]: This is worded weirdly. Maybe reword the second sentence better?

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+You can already see the way that this activity is being organized. Cards, checkpoints, and images goes in their own folder. Check out each of the folders and see how things are named.
+
+Cards are named from 1.md to \#.md.
+
+Checkpoints are named card\#-checkpoint.md.
+
+Images are named what the image is related to.
+
+Notice, the most important here is the README.md. The format of readmes have been documented here [https://about.bitproject.org/teams/engineering/readme-formats](https://about.bitproject.org/teams/engineering/readme-formats). Please follow the documentation and use the example in Github as a guide. Make sure you have all the information.
```
Comment (Jmendoza194) [inline teams/developer-relations/bible-for-developers/stage-4-breakdown.md:43]: What's the most important here? I think you should be more specific.  Either that, or rephrase your first sentence.

File: teams/developer-relations/bible-for-developers/general-items-breakdown.md (hunk 13 above)
Comment (joeyhuaa) [reply teams/developer-relations/bible-for-developers/general-items-breakdown.md:32]: thanks, i'll definitely take that into consideration!

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md (hunk 10 above)
Comment (joeyhuaa) [reply teams/developer-relations/bible-for-developers/stage-4-breakdown.md:53]: I agree with Atul, I think this section needs cleaning up, it reads pretty clunky. 
Also, suggestion for the first sentence: use "rewards" instead of "something"

Maybe don't try and explain the algorithm at all? Instead, just mention that there is an algorithm and developers will be told how to use it when they start... I feel like there is a bit too much info lumped together.

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md (hunk 18 above)
Comment (joeyhuaa) [reply teams/developer-relations/bible-for-developers/stage-4-breakdown.md:43]: Wording could be more precise. Also, maybe provide a picture example with just the README headings? That way, you omit info that would otherwise be there, simply for the purpose of providing an example outline.

File: teams/developer-relations/bible-for-developers/stage-4-breakdown.md (hunk 17 above)
Comment (joeyhuaa) [reply teams/developer-relations/bible-for-developers/stage-4-breakdown.md:19]: I think there could be a smoother transition from mentioning the LMS to folder structure walkthrough.

Also, I would advise actually typing out the folder structure (outline format). It is more work, but for someone reading this, it will make much more sense.

File: teams/developer-relations/bible-for-developers/README.md [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+{% page-ref page="stage-4-breakdown.md" %}
+
+## How This Checklist Is Used Beyond Development
+
+### Developer
+
+For Developers, this checklist will be used to help guide them in the production of curriculum. It will serve as a checklist to help determine what a developer should or should not do when they are creating labs and activities.
+
+For Example, as developers are doing the activity cards, there is a checklist box that states that there must be 2-3 checkpoints in the activity. By using this checklist as a guide, developers would then remember to add checkpoints somewhere in the list. This would then save the developer's reviewer's time as the reviewer would not have to ask the developer to add the missing content and wait again for the developer to implement what is missing.
```
Comment (joeyhuaa) [inline teams/developer-relations/bible-for-developers/README.md]: "Example" should be lower-case. Otherwise, looks good!

File: teams/developer-relations/bible-for-developers/README.md [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+### Manager
+
+For Managers, this guide will be useful for checking on the work of the developers. Managers can refer to this checklist page to see what the developers did wrong, making giving feedback easier than ever.
```
Comment (joeyhuaa) [inline teams/developer-relations/bible-for-developers/README.md]: The second sentence is a bit redundant.

I think it would be better to just say something like, "For Managers, this guide will be useful for checking on the work of the developers, streamlining the process of giving feedback."

File: teams/developer-relations/bible-for-developers/README.md [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+
+For example, In the following we have Michelle as a Manager/Reviewer. In the pull request Checklist portion of this page, there is a checkbox that states "Correct Markdown Formatting", and as seen below, that checkbox was put to good use. By checking in this page, it is now easier to spot what needs to be revised, leading to faster and more efficient feedback given to the developer.
+
+![](../../../.gitbook/assets/screen-shot-2020-03-30-at-8.51.29-pm.png)
```
Comment (joeyhuaa) [inline teams/developer-relations/bible-for-developers/README.md:96]: Some grammar/capitalization errors. 

Also, second sentence reads a bit awkward, could use better wording/sentence structure.
